Add msg port to control rotator's phase increment

igorauad opened the pull request:
Often receivers estimate the frequency offset on a block that is downstream with
respect to the frequency correction NCO. In such architectures, typically the
frequency offset estimator feeds the estimation back to the NCO.

This patch adds a message port to the rotator block that makes this architecture
feasible using the rotator as NCO. With that, a downstream block can estimate
frequency offset and send a message to the rotator in order to update its
rotating frequency (i.e. its phase increment).

The input message port receives a PMT pair containing both the new desired phase
increment as well as the absolute index of the rotator's output sample where
this new phase increment should start to take effect.

The rationale of scheduling the phase increment update to a specific absolute
index is as follows. Usually the receiver will rely on a frame structure and
will estimate the frequency offset based on pilots along the frame. If the true
frequency offset changes in the middle of the frame, the frame-averaged
frequency offset estimation likely becomes corrupted. It is often better to
update the frequency correction right at the start of a new frame. This way, the
true frequency offset is preserved in the course of a frame and only changes on
the boundary between consecutive frames.

For flexibility when using this feature, it is responsibility of the downstream
block (more generally the message sender) to schedule the update properly, based
on absolute index. Thus, for example, if there is a decimator in between the
rotator and the message sender, the sender will need to take the sample rate
conversion factor into account when computing the absolute index of the update.

mbr0wn commented:
@igorauad Do we have a CLA from you? I'm flagging this just in case we don't. If we do, many apologies in advance. I've been away from GR/github a bit so I'm not quite up to date.

What would definitely spice up this PR is a GRC-based example.

mbr0wn commented on gr-blocks/include/gnuradio/blocks/rotator_cc.h:
```diff
@@ -32,6 +32,18 @@ namespace blocks {
+ *
+ * phase_inc (input):
+ *
+ *    Receives a PMT pair with a new phase increment to be set on the rotator at
```
Suggestion: Make this a dict. Then the index can be optional. That could actually be useful for applications that don't have an actual control loop.

igorauad replied:
Done

mbr0wn commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -40,9 +40,17 @@ rotator_cc::sptr rotator_cc::make(double phase_inc)
 {
     set_phase_inc(phase_inc);
+
+    message_port_register_in(pmt::mp("phase_inc"));
```
We should make it a habit to not hard code the port name down here. Can you make a const further up?

igorauad replied:
Agreed. Updated.

mbr0wn commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -40,9 +40,17 @@ rotator_cc::sptr rotator_cc::make(double phase_inc)
+
+    message_port_register_in(pmt::mp("phase_inc"));
+    set_msg_handler(
+        pmt::mp("phase_inc"),
```
...and then reuse it here.

mbr0wn commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -52,6 +60,19 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
 
+void rotator_cc_impl::handle_phase_inc_msg(pmt::pmt_t msg)
+{
+    if (pmt::is_pair(msg)) {
```
What if not? Please log a warning or other debug message.

igorauad replied:
Do you prefer a print/warning message or really a `throw std::runtime_error`? I've put the latter, but let me know what you think.

mbr0wn replied:
I would have just logged the error and skipped the message, but I have no objective reason to say so.

mbr0wn commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -52,6 +60,19 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
+        if (pmt::is_uint64(offset) && pmt::is_real(phase_inc)) {
+            d_idx_next_inc_update = pmt::to_uint64(offset);
+            d_next_phase_inc      = pmt::to_double(phase_inc);
+            d_inc_update_pending  = true;
```
We'll need a queue of outstanding phase inc changes. Here, sending multiple messages will overwrite the earlier ones.

igorauad replied:
This is still missing. Will complete on my next spare time slot :)

igorauad replied:
I've just had a chance to add this. Used an `std::queue`. Let me know if OK.

igorauad commented:
Thanks for reviewing this PR.

> @igorauad Do we have a CLA from you? I'm flagging this just in case we don't. If we do, many apologies in advance. I've been away from GR/github a bit so I'm not quite up to date.

I don't have a CLA yet. I will send a message to Ben and will let you know.

> What would definitely spice up this PR is a GRC-based example.

I agree.  I wanted to propose the idea first, but I can definitely try to provide at least some unit tests, if not a GRC-based example. I'm working on an OOT module that would benefit from this and I thought this is really something for the in-tree rotator module, as it is such a common use case.

Also, something else that could be nice for this PR is tagging when the update is effectively handled by the rotator. I think this could be an optional behavior and would default to disabled. But it can be quite useful for some use cases. The main scenario I can think of is when the chain is something like:

rotator (as de-rotator) -> symbol sync / decimation -> frequency recovery

The frequency recovery block can schedule the update by taking the sample rate conversion (of the decimator) into account easily. In contrast, it may be more difficult (or at least more time consuming) to take the filter delay into account when scheduling the update. A solution is for the rotator to tag when it updates its increment and then for the frequency recovery block to check whether the update tag came where it expected. If it doesn't come where expected, the frequency recovery block can rely on the tag offset in order to infer the delay of the decimation block. 

This is something that we implemented in this [Blockstream Satellite - gr-blocksat module](https://github.com/Blockstream/gr-blocksat/blob/master/lib/frame_synchronizer_cc_impl.cc#L430). This is more specifically the consumer of the tag, which in turn schedules a frequency update on an upstream block that sits behind a polyphase clock sync (decimator) block.

Regarding your other comments, will work on them soon.

Thanks

mbr0wn commented:
I like the idea of optional tagging. Thanks for submitting a PR for discussion!

igorauad commented:
Also:
- I've added the mechanism for tagging the sample where the update starts to take effect.
- I've introduced some QA tests covering both the update that specifies sample offset as well as the update that does not.

igorauad commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -63,7 +102,38 @@ int rotator_cc_impl::work(int noutput_items,
+        d_idx_next_inc_update > n_written &&
+        d_idx_next_inc_update <= (n_written + noutput_items)) {
+
+        items_before_update = d_idx_next_inc_update - n_written - 1;
```
I've pushed this fix also. The `-1` is needed here in order for the new phase increment to start taking effect on sample index `d_idx_next_inc_update`.  I realised this after introducing the QA tests.

igorauad commented on gr-blocks/python/blocks/qa_rotator_cc.py:
```diff
@@ -0,0 +1,199 @@
+        self.tb.run()
+
+        for (x,y) in zip(self.sink.data(), expected_samples):
+            self.assertComplexAlmostEqual(x, y, places=5)
```
I am aware that there is `assertComplexTuplesAlmostEqual`. However, I couldn't get it to work. It would always pass even when it shouldn't. I couldn't understand why.

mbr0wn replied:
Meh. The macro is supposed to save you time developing, and it's clearly not, so let's move on.

igorauad commented on gr-blocks/python/blocks/qa_rotator_cc.py:
```diff
@@ -0,0 +1,199 @@
+                                pmt.intern("offset"),
+                                pmt.from_uint64(offset))
+
+        rot_ctrl = blocks.message_strobe(ctrl_msg, msg_period_ms)
```
This isn't very reliable and I accept suggestions. I'm using message period as 0 so that I try to get the message sent before processing the 100 samples that are tested. However, this does not work every time. Sometimes the samples are processed first. Hints?

igorauad replied:
@mbr0wn aside from this, I believe this MR is ready. If you have suggestions on how to solve the above testing issue, let me know.

To clarify, the goal here is to test a message sent to the rotator asking for an update of its's phase  increment. To do so, the message has to be sent first, before the rotator actually processes the samples. In this particular test the message asks for an update on sample offset 50, so `blocks.message_strobe` should convey the message before `blocks.rotator_cc` processes the first 50 samples, otherwise the phase increment will not be updated. This does not seem to work reliably using `blocks.message_strobe`, so not always the test will pass.

mbr0wn replied:
If you call `to_basic_block()._post()` to pass in the message instead of using a message strobe, you might be able to force the message into the queue before you start the FG.

mbr0wn replied:
Quick grep found this: https://github.com/gnuradio/gnuradio/blob/64ee803068269fe3b9563fd415c420fb513a1c9b/gr-digital/python/digital/qa_header_payload_demux.py#L332-L336

This might work for you also.

igorauad replied:
Thanks. I used `insert_tail`, which is what would `_post` would call AFAIU.

mbr0wn commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -32,17 +32,28 @@
+    message_port_register_in(port_id);
+    set_msg_handler(
+        port_id,
+        boost::bind(&rotator_cc_impl::handle_phase_inc_msg, this, _1));
```
```suggestion
        [this](pmt::pmt_t& msg){ this->handle_phase_inc_msg(msg); });
);
```

Please no boost::bind. And don't trust my lambda, I just winged it.

igorauad replied:
Fixed.

mbr0wn commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -40,6 +48,33 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
+                ? pmt::to_uint64(pmt::dict_ref(msg, offset_key, pmt::PMT_NIL))
+                : nitems_written(0) + 1; // update right away when not provided
+
+        d_inc_update_queue.push(std::make_pair(update_offset, phase_inc));
```
Almost! You  need to hold a mutex while accessing the queue, because the message handling and the work function are going to happen in separate threads. Since you're not already using it, using `d_setlock` is probably fine.

michaelld replied:
Here's the code, for both here and in `::work`:
```
gr::thread::scoped_lock guard(d_setlock);
```
Using the overall block lock like this is the way to go.

igorauad replied:
Add the `scoped_lock` to `handle_ctrl_msg()` and `work()` as suggested.

mbr0wn commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -51,7 +86,53 @@ int rotator_cc_impl::work(int noutput_items,
+
+    /* Find out if there is a phase increment update pending */
+    while (!d_inc_update_queue.empty() && !d_inc_update_pending) {
+        auto next_update = d_inc_update_queue.front();
```
You also need to hold the mutex for this line and the next.

michaelld commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -40,6 +48,33 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
 
+void rotator_cc_impl::handle_phase_inc_msg(pmt::pmt_t msg)
+{
+    const pmt::pmt_t inc_key = pmt::intern("inc");
```
these could be globally defined

michaelld replied:
for example `static const`, not just `const`

igorauad replied:
Done :+1:

mbr0wn commented:
@igorauad this is still a great change, but the mutex is really important. It would be great if you can finish the work you did here so we can merge this. If you're not available, I might close it just so we have a better overview of our PR queue, but of course you can reopen anytime you have changes. I'll give it a few days before doing that though.

igorauad commented:
@mbr0wn thanks. I haven't abandoned this change. However, I've been busy lately and haven't had a chance to complete the work. I'm thinking I might be able to work on it next week. This change is a requirement for an OOT module I'm developing, so I will definitely want to complete the work.

Please feel free to close the PR temporarily if that works best for your organization. I will reopen it as soon as I find a chance. That's all fine.

mbr0wn commented:
> Please feel free to close the PR temporarily if that works best for your organization. I will reopen it as soon as I find a chance. That's all fine.

Many thanks. I'll close it for the time being --  your code is still very much appreciated!

igorauad commented:
Hi @mbr0wn @michaelld ,

I would like to reopen this PR after a long time. I finally had a chance to get back to the project I was working on last year when I created this PR. So I've worked a bit more on it and ended up changing the implementation considerably. The essence is still the same, though. It will let other blocks control the phase increment adopted by the rotator block. 

I've rebased the changes using the new pybind11 structure, added the changes you had suggested last time, and a couple of extra unit test cases.

Let me know if you are available to take another look.

Thanks!

mbr0wn commented:
Just rebase and reopen. No process required! Or you can open a new PR. Up to you.

igorauad commented:
Thanks, @mbr0wn. The reopen button was disabled, but I figured out the problem after checking [this](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c). So I've just reopened it.

igorauad commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -40,18 +46,91 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
     d_r.set_phase_incr(exp(gr_complex(0, phase_inc)));
 }
 
+void rotator_cc_impl::handle_ctrl_msg(pmt::pmt_t msg)
+{
+    gr::thread::scoped_lock l(d_mutex);
+
+    static const pmt::pmt_t inc_key = pmt::intern("inc");
+    static const pmt::pmt_t offset_key = pmt::intern("offset");
+
+    if (!pmt::is_dict(msg)) {
+        throw std::runtime_error("rotator_cc: Rotator control message "
+                                 "must be a PMT dictionary");
+    }
+
+    if (!pmt::dict_has_key(msg, inc_key)) {
+        throw std::runtime_error("rotator_cc: Phase increment control "
+                                 "message must contain an \"inc\" key");
+    }
+
+    /* Define the new phase increment (mandatory) and the absolute sample offset
+     * when the update is to be applied (optional). If the latter is not defined
+     * in the control message, update the phase increment right away. */
+    phase_inc_update_t update{};
+    update.phase_inc = pmt::to_double(pmt::dict_ref(msg, inc_key, pmt::PMT_NIL));
+    update.offset = pmt::dict_has_key(msg, offset_key)
+                        ? pmt::to_uint64(pmt::dict_ref(msg, offset_key, pmt::PMT_NIL))
+                        : nitems_written(0);
+
+    d_inc_update_queue.push(update);
+}
+
```
In the previous implementation, the message port was named `phase_inc` and this method was called `handle_phase_inc_msg`. I've decided to make it more generic to facilitate future expansion. So I named the port `ctrl` and the message handler `handle_ctrl_msg`.

Not sure if this is common practice in the GR codebase, but the idea would be that we could send multiple commands to the rotator. This implementation adds a single command (to change the phase increment), but it would be easy to add other commands (e.g., to set the phase) using the same port and message handler function.

Now, two questions:
- Should we stick to this or should I revert back to a dedicated `phase_inc` port?
- If we use the suggested generic `ctrl` port, should I add a `command` field on the PMT message and call it `set_phase_inc`? The rationale is that, if other commands are added in the future, it would be easier to coordinate them with `command` field on the input `msg`.

willcode commented:
I think we should move toward having a single command/control message port on blocks, instead of adding new ports for each function. That way, if something wanted to command a frequency change and set the phase at the same time, it could. This also means that all keys would be optional. One key might still be required if another one is to make sense.

igorauad commented on gr-blocks/python/blocks/qa_rotator_cc.py:
```diff
@@ -0,0 +1,322 @@
+                                            expected_samples,
+                                            places=4)
+
+    def test_phase_inc_update_out_of_range(self):
+        """Test phase increment update sent for an out-of-range offset"""
+        self._setUp(n_samples=2**16)
+        n_half_samples = int(self.n_samples / 2)
+
+        # Schedule a phase increment update whose offset is initially
+        # out-of-range due to being in the future
+        new_phase_inc = 2 * np.pi * 0.1
+        self._post_phase_inc_ctrl_msg(new_phase_inc, offset=n_half_samples)
+
+        # Run the flowgraph and wait until the rotator block does some work
+        self.tb.start()
+        while (self.rotator_cc.nitems_written(0) == 0):
+            pass
+
+        # The out-of-range update (scheduled for the future) should not have
+        # been applied yet, but it should not have been discarded either.
+        self._assert_tags([], [])
+
+        # Wait until at least the first half of samples have been processed
+        while (self.rotator_cc.nitems_written(0) < n_half_samples):
+            pass
+
+        # Now, schedule an update that is out-of-range due to being in the past
+        self._post_phase_inc_ctrl_msg(new_phase_inc, offset=0)
+
+        # And run the flowgraph to completion
+        self.tb.wait()
+
+        # The increment update initially scheduled for the future should have
+        # been applied when processing the second half of samples. In contrast,
+        # the update scheduled for the past should have been discarded.
+        self._assert_tags([new_phase_inc], [n_half_samples])
```
I should note that this test may be slightly unreliable, given that it has a bit of racing. I couldn't think of a better way to implement it, though.

Essentially, I wanted to call the `_assert_tags` on L303 after the first `work` call on the rotator block. Also, I wanted to post a control message on L310 after processing half of the samples. So I added `while` loops, which are not perfect, but worked for me.

Let me know what you think.

igorauad commented:
> I think we should move toward having a single command/control message port on blocks, instead of adding new ports for each function. That way, if something wanted to command a frequency change and set the phase at the same time, it could. This also means that all keys would be optional. One key might still be required if another one is to make sense.

Cool, yes, that's what I had in mind. So, in this case, if `command == set_phase_inc`, the `inc` key is required. Otherwise, the `inc` is not mandatory. For instance, if there was a `set_phase` command, the required key would probably be named `phase`.

willcode commented:
Since there is already a `freq` port, that should be changed to be a `cmd` port now. The port would allow things like (making this up, may not all make sense):

`{'freq': 100000}` or `{'freq': 100000, 'phase': 0.1, 'offset': 10}` or `{'phase_inc': 0.1, 'offset': 10}`

with some rules about which combinations make sense.

igorauad commented:
> Since there is already a `freq` port, that should be changed to be a `cmd` port now. The port would allow things like (making this up, may not all make sense):
> 
> `{'freq': 100000}` or `{'freq': 100000, 'phase': 0.1, 'offset': 10}` or `{'phase_inc': 0.1, 'offset': 10}`
> 
> with some rules about which combinations make sense.
 
I see and I agree. We don't need a `cmd` key on the dictionary as long as we can infer the command based on the keys that are present. For example, if `phase_inc` is present, we should update the phase increment. If something else (e.g., `phase`) was present (say the message is `{'phase_inc': 0.1, 'offset': 10, 'phase': 0.2}`, we would handle both the phase and the phase increment. So I changed the implementation slightly in 9dc9dfe2fb1149d3fd979ab238fe4b42785c66ec.

willcode commented:
Yes, something like that. Then we can (later) add handling for freq to the same message handler, maybe keeping it also in the old message handler for compatibility.

We use 'cmd' in the signal source, and 'command' in the uhd blocks. Maybe one of those would be better than ctrl. This hasn't been well standardized. Someone else have an opinion on that?

willcode commented:
What is the purpose of having a message queue? From a technical perspective, it seems fine, but from a control loop perspective? Is there the thought that the desired phase shift will be known for future samples? Or is the thought that this will be used as a way to set up future configuration? I can see a potential use for setting up precise frequency hopping using something like this.

igorauad commented:
> What is the purpose of having a message queue? From a technical perspective, it seems fine, but from a control loop perspective? Is there the thought that the desired phase shift will be known for future samples? Or is the thought that this will be used as a way to set up future configuration? I can see a potential use for setting up precise frequency hopping using something like this.

I'm trying to address frequency correction on frame-by-frame processing schemes used in so many PHY implementations (e.g., DVB-S2). Usually, there is a frame structure with some pilot symbols at the beginning of the frame and then a pilot-aided carrier phase recovery loop that restarts on every frame. Whenever a new frame starts, and there are some pilots, the phase recovery loop will quickly snap back to the estimated carrier phase (more easily estimated with the pilots). Then, once the phase recovery loop starts to process the normal (non-pilot) data symbols, it will proceed with decision-directed carrier phase recovery (relying on decisions instead of know pilots), which is far more challenging under low SNR.

In addition to the carrier phase recovery, there is usually a separate frequency recovery block, which is meant to correct larger frequency errors (the so-called coarse freq. recovery). Furthermore, this coarse freq. recovery block usually comes first in the pipeline. In this context, the key point is that the coarse frequency correction block can disturb the carrier phase recovery block downstream in the pipeline. Any change in the frequency correction can easily disturb the phase recovery loop, especially if the frequency correction happens when the phase recovery loop is in its decision-directed processing operation (processing the data symbols instead of the pilots).

To overcome all this, it's quite important to update the coarse frequency correction only at the beginning of frames.  WIth frame-by-frame processing, the carrier phase recovery loop typically starts over from scratch on every frame. Hence, if the coarse frequency correction happens at the beginning of a frame, there is usually plenty of time for the phase recovery loop to adapt. As a result, the two blocks can coexist nicely.

So the queue is meant to provide deterministic scheduling of phase increment updates. The block sending the control message will say exactly when the update needs to be applied. Presumably, the sender block has the notion of frames already, so it knows exactly when a frame starts and schedules the update on that sample. 

There are some extra comments on the commit message (see b3f4232e12c21468783cf714423195f643cdb899).

willcode commented:
If you don't hear from anyone else, let's go with "cmd". The work "command" takes up too much space in GRC.

igorauad commented:
> If you don't hear from anyone else, let's go with "cmd". The work "command" takes up too much space in GRC.

Sounds good, @willcode. I've just renamed.

willcode commented:
I'll give it a try soon. By the way, the GRC and .h still have "ctrl".

igorauad commented:
> I'll give it a try soon. By the way, the GRC and .h still have "ctrl".

Ah, right! Fixed. Thanks

willcode commented:
Built/ran tests, and it went smoothly. The QA looks very complete, though I haven't spent the time to QA the QA.

One more naming thing ... once the tag flows downstream, `new_inc` could mean anything. I don't know what
a good name would be. Something like `phase_inc_update`? Something shorter but that still gives a clue what
the tag pertains to?

Other than that, I think we should stop torturing you and merge this :-)

igorauad commented:
> Built/ran tests, and it went smoothly. The QA looks very complete, though I haven't spent the time to QA the QA.
> 
> One more naming thing ... once the tag flows downstream, `new_inc` could mean anything. I don't know what
> a good name would be. Something like `phase_inc_update`? Something shorter but that still gives a clue what
> the tag pertains to? Just `rotator_update`?
> 
> Other than that, I think we should stop torturing you and merge this :-)

Thanks, @willcode No problem with code review rounds at all! 😄

I like the idea of a more descriptive tag. I think the tag should have a `rotator_` or `rot_` prefix so that downstream blocks can know where it comes from. Or, at least, it is easier to get a clue regarding which block is producing the tag if looking at a GUI plot. So, perhaps one of the following:

1. `rot_phase_inc`/`rotator_phase_inc`
2. `rot_inc_update`/`rotator_inc_update`
3. `rot_phase_inc_update`/`rotator_phase_inc_update`

The shortest one is `rot_phase_inc`. What do you think?

willcode commented:
rot_phase_inc sounds good.

igorauad commented:
> rot_phase_inc sounds good.

Ok 👍 Updated, squashed and rebased.

igorauad commented:
@willcode I've just pushed again to update the commit message with a `blocks: ` prefix.

mormj commented on gr-blocks/include/gnuradio/blocks/rotator_cc.h:
```diff
@@ -20,6 +20,23 @@ namespace blocks {
 /*!
  * \brief Complex rotator
  * \ingroup math_operators_blk
+ *
```
:+1: for adding documentation to the block

mormj commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -40,18 +46,94 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
     const gr_complex* in = (const gr_complex*)input_items[0];
     gr_complex* out = (gr_complex*)output_items[0];
 
 #if 0
```
This #if 0 code can go

igorauad replied:
Done (5f1052a5a3fd97e5c0a6b7e435aafcde1275ec99)

mormj commented on gr-blocks/lib/rotator_cc_impl.cc:
```diff
@@ -40,18 +46,94 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
+            if (d_tag_inc_updates) {
+                add_item_tag(0,
+                             next_update.offset,
+                             pmt::string_to_symbol("rot_phase_inc"),
```
The tag key can be stored as a static class variable so the pmt for it doesn't have to be regenerated each time

igorauad replied:
Agreed! Done in 11c3a1da1662bf13327782be721ccbcb4054db25

mormj reviewed (approved):
Some minor things, but overall looks good.

mormj commented:
I think it's also worth considering (beyond the scope of this PR) adding this functionality more broadly for many types of blocks - how do we consistently change a parameter value at a specific sample number, and indicate that the change has occurred.  This could apply to almost any block.

igorauad commented:
> I think it's also worth considering (beyond the scope of this PR) adding this functionality more broadly for many types of blocks - how do we consistently change a parameter value at a specific sample number, and indicate that the change has occurred. This could apply to almost any block.

That would be nice indeed. Also related is a general-purpose (and optionally registered) message port capable of calling public methods of the class with arguments provided via a PMT dictionary. The combination of such a message port with what you are suggesting (the ability to schedule changes at a specific sample offset) can enable lots of use cases. In the rotator case, if both features were available at a base class level, I could use the general-purpose port to call the pre-existing `set_phase_inc` method on a scheduled sample offset.

I'll see if I can understand how that could possibly be introduced at a base class level. As you said, it's something for a future investigation beyond the scope of this PR.

bistromath commented:
@mormj, what about your newsched work here?

https://github.com/gnuradio/newsched/pull/71

Seems like it would pretty seamlessly bridge that gap, although it might be out of the scope of this discussion.

bistromath commented:
Is this a candidate for its own GREP, or at least its own issue tag? A Qt-style signals/slots pattern for GRC would enable all sorts of fun applications.

mormj commented:
Certainly would be a great candidate for a GREP!

willcode commented:
There would be a lot to consider to do it right. Worth a read: https://github.com/pothosware/PothosCore/blob/master/include/Pothos/Managed/Class.hpp

Is there anything else about this PR that needs attention?

mormj commented:
I think this looks good as a solution for this block, so going ahead and merging.
